[FEATURE] Strict component blueprint support

Windvis opened the pull request:
This implements the blueprint work of [RFC #779](https://rfcs.emberjs.com/id/0779-first-class-component-templates/#blueprints).

I squashed all the work from #20511 into a single commit to make rebases easier.

Closes #20511

### Current status:
Should be good to review. I'll clean up the git history once I know if the current approach is fine and I no longer need the option to revert individual commits.

Commit:
Add loose/strict flags

Add failing test for strict flag

Add file name mapping for strict flag

Add gjs template blueprint

Filter files depending on loose/strict flag

Add gjs test fixture

fix component import path

Add component-test strict blueprint

filter blueprint files for component-test

Add strict component test test (no typo)

Add component test strict fixture

Use authoringFormat option instead of explicit flags

Use authoringFormat option instead of flags in component

Revert explicit fileMapTokens for strict format

Only log skippedJsFiles in loose mode

Don't store filteresJsFiles for authoring format

do not log skippedJsFiles in component-test either

fix afterInstall param location

Fix blueprint closing tags

Remove log

fix: add components to import path

Add yield to template-only gjs

Make component blueprint ts by default

Add js blueprints

Move test to outside of pod

remove default template test gjs

check rfc fixture

fix: nested component invocation

remove log

feat: add template tag aliasses

chore: rename param to component-authoring-format

fix: remove assert on test

chore: update test doc comments to reflect new test waiter api

chore: update doc help

Windvis committed:
Update to ember-cli v6.2

ember-cli is used in the tests, and some tests need the newer features to work as expected.
We also need a version that uses the `babel-remove-types` package that includes the fixes for .gts detyping to work.

Because of this update, some slight adjustments are needed to some pod related tests. The `usePods` option of ember-cli-blueprint-test-helpers' `setupPodConfig` util isn't compatible with newer ember-cli versions. As a workaround we use the `--pod` argument when generating files.

Windvis commented:
Something seems to go wrong when we remove the types from the new .gts test file.

```
  1) Blueprint: component-test
       in app
         with default setup
           component-test x-foo --strict:
     SyntaxError: unknown: Unexpected token, expected "," (12:46)

  10 |     // and update using state.myProperty = 1; await rerender();
  11 |     // Handle any actions with function myAction(val) { ... };/* ___NEWLINE___ */
> 12 |     await render(template = __TEMPLATE_TAG_0__;);/* ___NEWLINE___ */
     |                                               ^
  13 |     assert.dom().hasText('');/* ___NEWLINE___ */
  14 |     // Template block usage:
  15 |     await render(template = __TEMPLATE_TAG_1__;);/* ___NEWLINE___ */
  ```

I think this might be a bug in ember-cli's .gts [detyping logic](https://github.com/ember-cli/ember-cli/blob/8f8d02b56ffe201a98e33c8247b0569395f3d82d/lib/models/blueprint.js#L539-L566), but I'm not familiar enough with ember-template-imports to know the way forward here.

Windvis commented:
Cross posting the open question from https://github.com/emberjs/ember.js/pull/20511#issuecomment-2393872939:

In Octane the component-class defaults to --no-component-class. I'm wondering if that's the most useful default for .gts components though, especially since switching from template only .gts files to class based ones isn't as easy due to the types.

Since this is a new flag, and the RFC doesn't specify it, would it make sense to generate a class based .gts file by default, and a TO version if the --component-class=@ember/component/template-only is provided?

Windvis committed:
Patch ember-cli to fix a .gts detype problem

This applies the same changes as https://github.com/ember-cli/ember-cli/pull/10659 so we don't need to wait for a new ember-cli release.

Windvis commented on package.json:
```diff
@@ -127,7 +127,7 @@
     "dag-map": "^2.0.2",
     "decorator-transforms": "2.0.0",
-    "ember-cli": "^4.10.0",
+    "ember-cli": "^6.1.0",
```
This actually installed v6.2.2 due to the `^`so I should just list that version instead or use `~` instead.

ef4 commented:
Naming proposal from the tooling team meeting discussion: can we use the edition setting in package.json to control both component and route blueprint generation?

```js
 "ember": {
    "edition": "polaris" // causes all components and routes to generate with template tag
  },
```

Also, if separate settings for each feature are also desirable, the value could be "polaris" vs "octane"

```
// .ember-cli
componentAuthoringFormat = "polaris"
routeAuthoringFormat = "octane"
```

Windvis commented:
> Naming proposal from the tooling team meeting discussion: can we use the edition setting in package.json to control both component and route blueprint generation?

@ef4 I think we can use that edition value to flip the defaults, but I wouldn't use it as the only option. The [RFC explicitly mentioned](https://rfcs.emberjs.com/id/0779-first-class-component-templates/#transition-path) a `--strict` and `--loose` flag for the component  generator.

Checking the edition value feels a bit out of scope for this PR though, since it would require other code changes (or the merge of #20863). The wider ecosystem might also not be ready yet for a value other than `octane` so I don't think we can recommend people from updating that edition value in their package.json files until we actually are closer to shipping it?

My preference would be to stick with the `--component-authoring-format="strict"`  with `--strict` and `--loose` aliases, as currently implemented, get it merged, and follow up with a PR that adds support for setting `edition: "polaris"` where needed.

My question about the argument name wasn't for this PR, but the route one :smile:, but using `route-authoring-format` makes sense so I can go for that, and make it match the setup here.

Windvis committed:
Work work

Windvis commented on node-tests/fixtures/component-test/rfc232.gjs:
Probably should add a .gts version of this as well.

Windvis commented on node-tests/blueprints/component-test.js:
```diff
@@ -270,6 +271,52 @@ describe('Blueprint: component', function () {
+    });
   });
 
   describe('in addon', function () {
```
Probably should add some addon tests as well.

Windvis commented on node-tests/blueprints/component-test.js:
```diff
@@ -270,6 +271,52 @@ describe('Blueprint: component', function () {
+    it('component foo --strict', function () {
+      return emberGenerateDestroy(['component', 'foo', '--strict'], (_file) => {
+        expect(_file('app/components/foo.gjs')).to.equal(
+          fixture('component/template-only-component.gjs')
```
This one still fails because detyping the .gts version has a different output. We should probably use a .gjs without detyping for this scenario.

Windvis commented:
The smoke test scenarios seem to fail because of the included patch for ember-cli:

```
ERR_PNPM_PATCH_NOT_APPLIED  The following patches were not applied: ember-cli@6.2.2

Either remove them from "patchedDependencies" or update them to match packages in your dependencies.
Error: Process completed with exit code 1.
```

Not sure why though, since that patch config is not in the package.json of the smoke test app. Not sure what the fix is here. I think we _need_ the patch, unless the fix in ember-cli is backported and released, and we can update to that one.

Windvis commented on node-tests/fixtures/component/template-only-component.gts:
```diff
@@ -0,0 +1,16 @@
+  Element: null;
+}
+
+<template>
```
This needs to be `export default <template>` it seems, otherwise it doesn't seem to work. Not sure if that's a bug or intended.

ef4 replied:
I think that is a typescript or glint issue. Template tag by itself is supposed to imply export default, and that definitely works when no TS is involved.

NullVoxPopuli replied:
I think it's a babel-plugin-ember-template-compilation issue, actually.

Here is a failing test: https://github.com/emberjs/babel-plugin-ember-template-compilation/pull/70

NullVoxPopuli replied:
from discord: https://discord.com/channels/480462759797063690/485447409296736276/1337849624635113593

NullVoxPopuli replied:
maybe another failing test? https://github.com/embroider-build/content-tag/pull/96

idk, this one seems like it's barfing on TS Syntax for some reason?

johanrd replied:
Yes, I also see this when migrating from ember-route-template. The workaround to add `export default` before `<template>` seems to work, but it was unexpected per the [examples in the RFC](https://rfcs.emberjs.com/id/1046-template-tag-in-routes/) that were linked to in the release notes of 6.3.0.

Windvis replied:
Resolved by content-tag v3.1.2 so we can safely output it like this :+1:

Windvis committed:
Update the component-test tests

- add gts
- make the gjs consistent.

Windvis committed:
Add some more tests

Windvis committed:
fixup tests

Windvis committed:
Update to ember-cli 6.3.0-beta.3

6.3.0 hasn't been released yet, but we need the content-tag v3 dependency to fix an issue when detyping the blueprint files.

Windvis committed:
don't depend on .gts type removal

The .gts type removal is buggy at the moment, and causes this PR to be blocked. We now add .gjs blueprints and strip the version we don't need, which bypasses the detyping problems.

Windvis committed:
Remove the ember-cli-patch

We no longer need it since we don't depend on the type removal feature.

Windvis committed:
fixup stop detyping

Windvis committed:
Remove the blueprint-js files

That folder is no longer used.

Windvis commented on blueprints/component/index.js:
```diff
@@ -135,6 +161,21 @@ module.exports = {
     }
+    if (this.options.componentAuthoringFormat === 'strict') {
+      const strictFilesToRemove =
+        this.options.isTypeScriptProject || this.options.typescript ? '.gjs' : '.gts';
```
I'm not 100% sure if the `isTypeScriptProject` check is needed. It's possible that when that is true, `options.typescript` is also true, but I would have to double check.

Windvis committed:
rename the fixtures

This is in line with the PR that cleans up ember-cli-qunit tests

Windvis commented:
I decided to work around all the different .gts detyping issues by simply adding .gjs blueprint files and selecting the correct one based on the host project. This allows us to drop all the workarounds and also means we are no longer blocked by the ember-cli backport PR.

I think this is the best way forward, and we can always refactor later if needed.

Windvis commented on node-tests/fixtures/component-test/addon.gts:
```diff
@@ -0,0 +1,28 @@
+import { module, test } from 'qunit';
+import { setupRenderingTest } from 'dummy/tests/helpers';
+import { render } from '@ember/test-helpers';
+import Foo from 'dummy/components/foo';
```
I'm not sure about this one, I think we may want to import from the addon namespace in this scenario, right? So from `my-addon/components/foo` in the tests?

ef4 replied:
Yes, I think it would be better to use the addon's namespace.

Windvis committed:
Update the component import path in addon tests

We now use the addon name instead of the modulePrefix.

Windvis committed:
Remove the `unit` test type for strict components

Windvis committed:
fix lint issue

it seems I messed up during the rebase.

Windvis committed:
fix attempt 2.. mondays <3

kategengler commented on package.json:
```diff
@@ -127,7 +127,7 @@
     "dag-map": "^2.0.2",
     "decorator-transforms": "2.0.0",
-    "ember-cli": "^4.10.0",
+    "ember-cli": "^6.3.0-beta.3",
```
Should this now go to 6.3.0?

kategengler replied:
I'm going to make this change

kategengler commented:
I'm happy to merge after updating the cli version to non-beta.

kategengler committed:
Update to published ember-cli 6.3
